Unify background placement code

pyfisch opened the pull request:
Merges the implementations for background-image placement
from gradients and images. Add missing parts and fix bugs.

Now supported are the CSS properties:

* background-attachment (except for local value, see #19650)
* background-clip
* background-origin
* background-position-x/y
* background-repeat
* background-size

It should be noted that backgrounds are not clipped to
rounded border corners.
(This was done before but worked only in simple cases)
See: #19649

This solves the following issues:
closes #19626
closes #16657
closes #19482 (examples from http://lea.verou.me/css3patterns/ are rendered perfectly but the round border is completely ignored now)
closes #19577

- `./mach build -d` does not report any errors
- `./mach test-tidy` does not report any errors

I enabled a few tests with the first commit but I have written about a dozen manual tests I will try to turn into ref tests either before or after this patch lands.

@bors-servo try

The relationship between the different inputs is visualized in this flowchart:
![flowchart-background](https://user-images.githubusercontent.com/2781017/34394430-5a06c72c-eb59-11e7-9d51-3d23e2215f07.png)


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19651)
<!-- Reviewable:end -->

pyfisch committed:
Unify background placement code

Merges the implementations for background-image placement
from gradients and images. Add missing parts and fix bugs.

Now supported are the CSS properties:

* background-attachment (except for local value)
* background-clip
* background-origin
* background-position-x/y
* background-repeat
* background-size

It should be noted that backgrounds are not clipped to
rounded border corners.

highfive commented:
Heads up! This PR modifies the following files:
 * @emilio: components/layout/display_list_builder.rs, components/layout/fragment.rs, components/layout/webrender_helpers.rs

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1024,6 +1041,176 @@ fn convert_ellipse_size_keyword(keyword: ShapeExtent,
+    // normal but only display a single tile.
+    if image_size * 2 >= *size {
+        tile_image(position,
+                                        size,
```
nit: Indentation here is off.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1184,172 +1362,103 @@ impl FragmentDisplayListBuilding for Fragment {
-                            bounds.size.height)
-            }
+    fn compute_background_placement(&self,
+                                    state: &mut DisplayListBuildState,
```
nit: Maybe indenting these using block indentation, like:

```rust
fn compute_background_placement(
    &self,
    state: &mut ...,
) -> (Rect<Au>, Size2D<Au>, Size2D<Au>, Rect<Au>) {
```

is nicer. We should reformat this file to that style eventually anyway.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1184,172 +1362,103 @@ impl FragmentDisplayListBuilding for Fragment {
+                                    absolute_bounds: Rect<Au>,
+                                    intrinsic_size: Option<Size2D<Au>>,
+                                    index: usize)
+                                    -> (Rect<Au>,
```
nit: Maybe a struct would be nicer instead of a tuple? It's not clear what each field means without looking at the function implementation.

emilio reviewed (commented):
This looks like a pretty neat cleanup!

Haven't gone through the math yet, but looks nice overall.

pyfisch commented:
Fixed nits.

pyfisch committed:
Test with fixes for background repeat: spaced

Test the interaction between background-repeat: spaced
and CSS borders. Fix tile_image_axis() function.

pyfisch commented:
Another bug is that `background-attachment: fixed` is scrolled with the window contents while it should stay attached to the viewport. (This was *not* broken by this PR) See these two tests:
* servo/webrender#1301 https://staktrace.github.io/moz-pages/bug/1368496/
* servo/tests/wpt/web-platform-tests/css/css-backgrounds/background-attachment-fixed.html

Can you give me an example how to accomplish the fixed positioning? The webrender issue mentions push_clip_and_scroll_info but how exactly is it used here?

emilio commented:
> Can you give me an example how to accomplish the fixed positioning? The webrender issue mentions push_clip_and_scroll_info but how exactly is it used here?

Probably @glennw or @kvark can :)

kvark commented:
Also cc @mrobinson

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1024,6 +1028,194 @@ fn convert_ellipse_size_keyword(keyword: ShapeExtent,
+              absolute_anchor_origin: Au,
+              image_size: Au) {
+    // Avoid division by zero below!
+    if image_size == Au(0) {
```
I think it'd make sense to comment why the outparams are sane in this case, or why it doesn't matter..

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1024,6 +1028,194 @@ fn convert_ellipse_size_keyword(keyword: ShapeExtent,
+    let own_size = if let Some(size) = intrinsic_size {
+        size
+    } else {
+        return match bg_size {
```
This may be cleaner with `match intrinsic_size { ...`, but no strong preference I guess

pyfisch replied:
I am not sure if understand this. `match bg_size { ...` matches on the CSS property `background-size`. Maybe you meant I should replace `own_size` with `intrinsic_size`?

mbrubeck replied:
As a replacement for `if let Some(size) = intrinsic_size` on line 1050

pyfisch replied:
Thanks. Will change it.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1184,172 +1367,100 @@ impl FragmentDisplayListBuilding for Fragment {
+            &mut tile_spacing.height,
+            pos_y,
+            css_clip.origin.y,
+            css_clip.size.height);
```
nit: Paren on the next line, here and in the rest of the calls using this format.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1184,172 +1367,100 @@ impl FragmentDisplayListBuilding for Fragment {
+            css_clip.origin.y,
+            css_clip.size.height);
+
+        return BackgroundPlacement { bounds, tile_size, tile_spacing, css_clip }
```
nit: No need for `return`.

emilio reviewed (commented):
This looks fine to me. Not sure if you want to wait for the fixed image stuff or land this before, or prefer review from @glennw / @kvark / @mrobinson.

pyfisch commented:
Force pushed the fixed nits.

pyfisch commented:
> This looks fine to me. Not sure if you want to wait for the fixed image stuff or land this before, or prefer review from @glennw / @kvark / @mrobinson.

I'd like to land this as is.

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1024,6 +1028,201 @@ fn convert_ellipse_size_keyword(keyword: ShapeExtent,
+                                 -> Size2D<Au> {
+    match intrinsic_size {
+        None => {
+            match bg_size {
```
Well, I would've keep the early return, just using `match`, but it's not a big deal, this looks fine too, thanks!

emilio commented on components/layout/display_list_builder.rs:
```diff
@@ -1024,6 +1028,201 @@ fn convert_ellipse_size_keyword(keyword: ShapeExtent,
+    // Images with a zero width or height are not displayed.
+    // Therefore the positions do not matter and can be left unchanged.
+    // NOTE: A possible optimization is not to build
+    // display lists in this case at all.
```
nit: `s/display lists/display items/`

emilio reviewed (approved):
Ok, let's land it as soon as that nit is addressed :)

emilio commented:
Mind filing issues for the remaining bits @pyfisch? Thanks a lot!

pyfisch commented:
> nit: s/display lists/display items/

done.

I've opened #19650, #19669, #19672 for the remaining bits. Anything missing?

pyfisch commented:
Travis failure is: `oci runtime error: exec failed: container_linux.go:265: starting container process caused "could not create session key: disk quota exceeded"`

emilio commented:
Yeah, same failure happened over #19668. We don't gate on travis though, so should be gtg, thanks!

@bors-servo r+

bors-servo commented:
:pushpin: Commit 9849295 has been approved by `emilio`

<!-- @bors-servo r=emilio 984929559f037d90d7e39f9889d8ed146f88e4c9 -->

bors-servo commented:
:hourglass: Testing commit 984929559f037d90d7e39f9889d8ed146f88e4c9 with merge f6111c52e0afadb47f6f0a662cb45d89a109c06e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7398)

pyfisch commented:
Sorry for the failing tests. My computer is really slow at running them all. Are there options to speed up `./mach test-wpt`? I just saw that the builders run  `test-wpt --release --processes 24 --total-chunks 2 --this-chunk 1 --always-succeed`. Are the tests much faster with release builds and what number of processes should be used on a quad-core Laptop? Are the other parameters relevant?

At some point I was allowed to start bors with

@bors-servo try

but this doesn't work anymore. (Or does it?)

bors-servo commented:
:hourglass: Trying commit cf12e27364c8fe29d07ba3ba011f856c344546b8 with merge 62d21791855a6989de6cd8bbcad567e3d7cc7ceb...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3106)

emilio commented:
It does! And those tests are unexpected passes, so we need to mark them passing, not fixing anything :)

pyfisch commented:
> so we need to mark them passing, not fixing anything :)

Bad word choice on my side. I meant to mark them as passing by deleting the metadata files.

pyfisch commented:
All tests passed. (Except on Mac where some builders failed.)

@emilio Can you please approve?

emilio commented:
Sure thing! Thanks a lot @pyfisch!

@bors-servo r+

bors-servo commented:
:pushpin: Commit cf12e27 has been approved by `emilio`

<!-- @bors-servo r=emilio cf12e27364c8fe29d07ba3ba011f856c344546b8 -->

bors-servo commented:
:hourglass: Testing commit cf12e27364c8fe29d07ba3ba011f856c344546b8 with merge 691bff86b6f57da37c2e0af99aa7f72fc1b4f9d6...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10246), [arm32](http://build.servo.org/builders/arm32/builds/10482), [arm64](http://build.servo.org/builders/arm64/builds/10510), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10536), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7386), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7401), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10606), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/3024), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3039), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7288), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7348), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3107), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3061), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5903)
Approved by: emilio
Pushing 691bff86b6f57da37c2e0af99aa7f72fc1b4f9d6 to master...

CYBAI commented:
@pyfisch Congrats!!!!!!
